File: src/Ej3/Razas/PersonajeBuilder.java
package Razas;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;

import Clases.ClaseBuilder;

public abstract class PersonajeBuilder {
	protected Personaje personaje;
	protected ClaseBuilder claseBuilder;
	static protected String CFG_PATH = "src/Ej3/cfg/raza/";
	protected String raza;

	public PersonajeBuilder(ClaseBuilder claseBuilder) {
		this.claseBuilder = claseBuilder;
		personaje = new Personaje();
	}

	public void CrearPersonaje() {
		System.out.println("Creando personaje: "+raza);
		CrearAtributos();
	};

	public void CrearAtributos() {
		CrearPrimarios();
		CrearSecundarios();
		personaje.SetClass(claseBuilder.GetClassName());
		personaje.SetRace(raza);
	};

	public void CrearPrimarios() {
		var attr = LoadCfg();
		personaje.SetPrimaryAttributes(attr);
	};

	/* Es comun a todas las clases */
	private HashMap<String, Integer> LoadCfg() {
		HashMap<String, Integer> temp = new HashMap<>();
		// Leer el archivo y construir el HashMap
		try (BufferedReader br = new BufferedReader(new FileReader(PersonajeBuilder.CFG_PATH + raza+".txt"))) {
			String line;
			while ((line = br.readLine()) != null) {
				// Dividir la línea en clave y valor usando ":"
				String[] parts = line.split(":");
				if (parts.length == 2) {
					String key = parts[0].trim(); // Clave sin espacios
					int value = Integer.parseInt(parts[1].trim()); // Valor como entero
					temp.put(key, value); // Agregar al HashMap
				} else {
					System.out.println("Error: línea no válida - " + line);
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
		return temp;
	}

	public void CrearSecundarios() {
		claseBuilder.CrearClase(personaje.GetPrimary());
		personaje.SetSecondaryAttributes(claseBuilder.GetClase());
	}

	public Personaje GetPersonaje(){
		return personaje;
	}
}

File: Practica 2/ej4_creador_personajes/lib/Razas/PersonajeBuilder.dart

import 'dart:collection';
import 'dart:io';
import 'dart:convert';

import 'package:ej4_creador_personajes/Razas/Personaje.dart';
import 'package:ej4_creador_personajes/Clases/clasesbuilder.dart';

abstract class PersonajeBuilder{
  late Personaje personaje;
  late ClaseBuilder claseBuilder;
  static String CFG_PATH = '../../cfg';
  late String raza;  

  PersonajeBuilder(ClaseBuilder claseBuilder){
    this.claseBuilder = claseBuilder;
    personaje = Personaje();
  }

  HashMap<String, double> loadCFG() {
    HashMap<String, double> temp = HashMap<String, double>();

    // Obtener la ruta del archivo de configuración correspondiente a la clase
    String cfgFilePath = '${ClaseBuilder.CFG_PATH}/raza/$raza.txt';
    // Verificar si el archivo existe
    File cfgFile = File(cfgFilePath);
    if (!cfgFile.existsSync()) {
      print("El archivo de configuración $cfgFilePath no existe.");
      return temp;
    }

    // Leer el archivo de configuración línea por línea
    List<String> lines = cfgFile.readAsLinesSync();
    for (String line in lines) {
      List<String> parts = line.split(':');
      if (parts.length == 2) {
        String attributeName = parts[0].trim();
        double attributeValue = double.tryParse(parts[1].trim()) ?? 0;  // Manejar posibles errores de conversión
        temp[attributeName] = attributeValue;
      }
    }

    return temp;
  }

  void crearPersonaje(){
    crearAtributos();
  }

  void crearAtributos(){
    crearPrimarios();
    crearSecundarios();
    personaje.SetClass(claseBuilder.getClassName());
    personaje.SetRace(raza);
  }

  void crearPrimarios(){
    var attr = loadCFG();
    personaje.SetPrimaryAttributes(attr);
  }

  void crearSecundarios(){
    claseBuilder.crearClase(personaje.GetPrimary());
    personaje.SetSecondaryAttributes(claseBuilder.getClase());
  }

  Personaje getPersonaje(){
    return personaje;
  }

}